Expand FromX support and add route vs. query param inference

captainsafia opened the pull request:
Part of https://github.com/dotnet/aspnetcore/issues/46275.

- Add support for parameters with `FromRoute` and `FromHeader` attribute
  - Includes support for `Name` argument on attributes
  - Includes support for requiredness checks via nullability and default value checks
- Add support  for `Name` argument in `FromQuery` attribute
- Add support for default value-based requiredness check on `FromQuery` attributes
- Update tests to use `SharedTypes` instead of types in test source
- Add support for inference for parameters between route and query values

captainsafia committed:
Expand FromX support and add route vs. query param inference

captainsafia committed:
Fix incrementality test

captainsafia commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs:
```diff
@@ -9,11 +9,11 @@ internal static class EndpointParameterEmitter
     {
         return $"""
-                        var {endpointParameter.Name}_local = {endpointParameter.AssigningCode};
+                        var {endpointParameter.HandlerArgument} = {endpointParameter.AssigningCode};
```
I was actually thinking the other way around. `EndpointParameter` should be strictly about processing syntax to set state on the parameter objects and `EndpointParameterEmitter` should handle constructing the blocks of code.

I'll add a commit to showcase what this might look like.

captainsafia committed:
Move code emissions out of parsing logic

captainsafia commented:
@dotnet/minimal-apis Can I get a review? 🙏🏽

eerhardt commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/SharedTypes.cs:
```diff
@@ -1,8 +1,29 @@
+    public bool IsComplete { get; set; }
+}
+
+public interface ITodo
```
Is this type used? I don't see any references to it.

captainsafia replied:
This can be removed for now.

eerhardt commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorIncrementalityTests.cs:
```diff
@@ -44,8 +45,12 @@ public async Task MapAction_ChangeReturnType_TriggersUpdate()
-        var updatedSource = @"app.MapGet(""/"", ([FromBody] Todo? todo) => TypedResults.Ok(todo));";
+        var source = $"""app.MapGet("/", ([{typeof(FromBodyAttribute)}] {typeof(Todo)} todo) => TypedResults.Ok(todo));""";
+        var updatedSource = $"""
+#pragma warning disable CS8622
```
For my learning, why do we get this warning here?

captainsafia replied:
This is a result of the bug outlined in https://github.com/dotnet/aspnetcore/issues/46622

eerhardt commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs:
```diff
@@ -35,75 +37,58 @@ public async Task MapAction_NoParam_StringReturn(string source, string httpMetho
+        get
+        {
+            var expectedBody = "TestQueryValue";
+            var fromQueryRequiredSource = $"""app.MapGet("/", ([{typeof(FromQueryAttribute)}] string queryValue) => queryValue);""";
```
Do we really need to use `typeof` here? I think it is clutter and confusing. It isn't like we are ever going to rename `FromQueryAttribute`.

```suggestion
            var fromQueryRequiredSource = """app.MapGet("/", ([FromQuery] string queryValue) => queryValue);""";
```

captainsafia replied:
Although we need this for the `Todo` type as mentioned for the reasons below, we don't strictly need it for `FromQuery` type since that's only defined in MVC. The main benefit we gain here is not having to have the `using` for the attributes in the test source.

eerhardt commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs:
```diff
@@ -291,13 +276,18 @@ public async Task MapAction_NoParam_AnyReturn(string source, string expectedBody
-object GetTodo() => new Todo() { Name = "Test Item"};
+    public static IEnumerable<object[]> MapAction_NoParam_ComplexReturn_Data => new List<object[]>()
+    {
+        new object[] { $$"""app.MapGet("/", () => new {{typeof(Todo)}}() { Name = "Test Item"});""" },
```
Similar comment here about needing to use `typeof`. If `Todo` was ever renamed, this code would fail to compile and the test would fail. I think it would be safe (and more readable) to just hard-code `Todo` here.

```suggestion
        new object[] { """app.MapGet("/", () => new Todo() { Name = "Test Item"});""" },
```

captainsafia replied:
We need to use the same `Todo` type in the compiled code that exists in the test code. Previously, the test code had its own `Todo` type defined that happened to have the same properties as the `Todo` type in the test code with differing nullability.

This is incorrect because we end up instantiated one type in test code and using another in the actual endpoint. Using the same type is more consistent and doesn't require us to rely on properties incidentally being deserialized the same way.

eerhardt commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitSource_SimpleReturn_Snapshot.generated.txt:
```diff
@@ -0,0 +1,1804 @@
+                    {
+                        var wasParamCheckFailure = false;
+                        // Endpoint Parameter: value (Type = string?, IsOptional = True, Source = RouteOrQuery)
+                        var value_raw = GeneratedRouteBuilderExtensionsCore.ResolveFromRouteOrQuery(httpContext, "value", options?.RouteParameterNames);
+                        var value_local = value_raw;
```
Why do we need `value_local` here?

captainsafia replied:
This is meant to indicate where the `TryParse` code in https://github.com/dotnet/aspnetcore/pull/46696 would go. It mostly exists to make rebasing easier since we'll need to support `TryParse` on parameters from multiple sources.

eerhardt commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitSource_SimpleReturn_Snapshot.generated.txt:
```diff
@@ -0,0 +1,1804 @@
+                        var value_raw = GeneratedRouteBuilderExtensionsCore.ResolveFromRouteOrQuery(httpContext, "value", options?.RouteParameterNames);
+                        var value_local = value_raw;
+
+                        if (wasParamCheckFailure)
```
When can `wasParamCheckFailure` ever be `true`?

captainsafia replied:
Good catch! It would be `true` if the parameter was required and wasn't provided in either the route or the query. I'll add a test case for this.

eerhardt commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:
```diff
@@ -118,25 +128,16 @@ private static bool TryGetSpecialTypeAssigningCode(ITypeSymbol type, WellKnownTy
+            return false;
         }
-        return false;
+        isOptional |= fromBodyAttribute.TryGetNamedArgumentValue<bool>("AllowEmpty", out var allowEmptyValue) && allowEmptyValue;
```
How does this work with something like the Mvc `AllowEmpty` property?

https://github.com/dotnet/aspnetcore/blob/ce8e0ce086b0be93d25b985e1c4cdf3fe6d7bc21/src/Mvc/Mvc.Core/src/FromBodyAttribute.cs#L26-L30

My understanding is that the attribute doesn't have a ctor parameter, and instead has this _other_ parameter `EmptyBodyBehavior` that affects the `AllowEmpty` property (which implements the interface explicitly).

It would seem that in order for the Mvc FromBody attribute to work, you need an actual instance of the attribute to call the `AllowEmpty` property on.

captainsafia replied:
https://github.com/dotnet/aspnetcore/issues/46639

I think this should actually be easy to do in this PR...

eerhardt commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs:
```diff
@@ -87,15 +153,31 @@ internal static string EmitServiceParameterPreparation(this EndpointParameter en
+            $"httpContext.RequestServices.GetRequiredService<{endpointParameter.Type}>()";
+
         // Requiredness checks for services are handled by the distinction
         // between GetRequiredService and GetService in the AssigningCode.
```
```suggestion
        // between GetRequiredService and GetService in the assigningCode.
```

Actually, is this comment even necessary anymore?

captainsafia replied:
> Actually, is this comment even necessary anymore?

Ish. It would be helpful to clarify here that requiredness checks for services will result in an exception being thrown (by virtue of `GetRequiredService` being called) not the 400 status code behavior that exists for other types.

eerhardt commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs:
```diff
@@ -87,15 +153,31 @@ internal static string EmitServiceParameterPreparation(this EndpointParameter en
                         {endpointParameter.EmitParameterDiagnosticComment()}
```
(nit on existing code)

Does this need to be wrapped in a `"""` string? Why can't it just be:

```C#
builder.AppendLine(endpointParameter.EmitParameterDiagnosticComment());
```

captainsafia replied:
Indentation! 😢 

But see https://github.com/dotnet/aspnetcore/pull/46845 for some goodies.

captainsafia commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:
```diff
@@ -299,6 +299,13 @@ namespace Microsoft.AspNetCore.Http.Generated
             return (false, default);
         }
+
+        private static StringValues ResolveFromRouteOrQuery(HttpContext httpContext, string parameterName, IEnumerable<string>? routeParameterNames)
```
Ah, interesting point. It doesn't get used in this particular test scenario. We can probably do something that only emits the `ResolveFromRouteOrQuery` method if we ended up doing route-vs-query inference in the generated code.

We would have to do some bookkeeping to track when that has happened.

Alternatively, it might be easier to just inline this resolution check.

captainsafia committed:
Address feedback from peer review

- Add support for EmpyBodyBehavior check
- Update source snapshots to implicit tests
- Avoid codegenning `TryResolveRouteOrQuery` if not needed
- Clean up test types

captainsafia committed:
Remove string interpolation in source under test

eerhardt commented on src/Shared/RoslynUtils/SymbolExtensions.cs:
```diff
@@ -82,4 +82,25 @@ public static ImmutableArray<IParameterSymbol> GetParameters(this ISymbol? symbo
+            if (namedArgument.Key == argumentName)
+            {
+                var routeParameterNameConstant = namedArgument.Value;
+                argumentValue = (T)routeParameterNameConstant.Value!;
```
What if the argument is set to `null`? Are we sure this will always be non null?

captainsafia replied:
Ah, good call! Yes, it's possible to do something like `[FromQuery(Name = null)]` for example. I updated the implementation here and added a guard against this scenario.

eerhardt commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:
```diff
@@ -113,8 +113,8 @@ namespace Microsoft.AspNetCore.Http.Generated
-                        // Endpoint Parameter: todo (Type = TestMapActions.Todo, IsOptional = False, Source = JsonBody)
-                        var (isSuccessful, todo_local) = await GeneratedRouteBuilderExtensionsCore.TryResolveBody<TestMapActions.Todo>(httpContext, false);
+                        // Endpoint Parameter: todo (Type = Microsoft.AspNetCore.Http.Generators.Tests.Todo, IsOptional = False, Source = JsonBody)
+                        var (isSuccessful, todo_local) = await GeneratedRouteBuilderExtensionsCore.TryResolveBody<Microsoft.AspNetCore.Http.Generators.Tests.Todo>(httpContext, false);
```
```suggestion
                        var (isSuccessful, todo_local) = await GeneratedRouteBuilderExtensionsCore.TryResolveBody<global::Microsoft.AspNetCore.Http.Generators.Tests.Todo>(httpContext, false);
```

eerhardt commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:
```diff
@@ -143,24 +143,24 @@ namespace Microsoft.AspNetCore.Http.Generated
                             httpContext.Response.StatusCode = 400;
                         }
-                        var result = await filteredInvocation(new EndpointFilterInvocationContext<TestMapActions.Todo>(httpContext, todo_local!));
+                        var result = await filteredInvocation(new EndpointFilterInvocationContext<Microsoft.AspNetCore.Http.Generators.Tests.Todo>(httpContext, todo_local!));
```
```suggestion
                        var result = await filteredInvocation(new EndpointFilterInvocationContext<global::Microsoft.AspNetCore.Http.Generators.Tests.Todo>(httpContext, todo_local!));
```

eerhardt commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:
```diff
@@ -60,19 +77,12 @@ public EndpointParameter(IParameterSymbol parameter, WellKnownTypes wellKnownTyp
     public EndpointParameterSource Source { get; }
 
-    // TODO: If the parameter has [FromRoute("AnotherName")] or similar, prefer that.
+    // Omly used for SpecialType parameters that need
```
```suggestion
    // Only used for SpecialType parameters that need
```

eerhardt commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:
```diff
@@ -118,25 +128,16 @@ private static bool TryGetSpecialTypeAssigningCode(ITypeSymbol type, WellKnownTy
+            return false;
         }
-        return false;
+        isOptional |= fromBodyAttribute.TryGetNamedArgumentValue<int>("EmptyBodyBehavior", out var emptyBodyBehaviorValue) && emptyBodyBehaviorValue == 1;
```
Does this code still handle a custom IFromBodyMetadata attribute when the `AllowEmpty` property is set in the attribute?

captainsafia replied:
I was a little iffy on whether or not we should support both. We use the custom `FromBodyAttribute` implementation in our RDF tests ([ref](https://github.com/dotnet/aspnetcore/blob/a11fec5fad546a315b4565af534f9efccec010b9/src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs#L7804-L7807)) but I don't think this is a common pattern in user code.

I ended up adding it back since we're not in a position to support `FromBody` vs. `FromService` inference and might need to implement a `FromBody` attribute in our own source to avoid importing the Mvc namespace.

eerhardt reviewed (approved):
LGTM. I just had a few more minor comments.

captainsafia committed:
Address review feedback

- Guard parameter name from attribute against null values
- Support FromBody with AllowEmpty named argument

captainsafia committed:
Undo format change

captainsafia commented:
/azp run
